docs/specs/user-profile/design.md

```diff
@@ -0,0 +1,313 @@
+It is subject to [automatic population](#automatic-population).
+
+### family\_name
+
+The family name or the last name of the end-user.
+The input control is a simple text input without validation.
+It is subject to [automatic population](#automatic-population).
+
+## Automatic population
+
+> Automatic population CANNOT be turned off!
+
+Some of the standard attributes are subject to automatic population.
+
+The automatic population runs when one of the following situation happens:
+
+- When the end-user signs up.
+- When the end-user adds an identity.
+- When the end-user updates an identity.
+- When the end-user removes an identity.
+
+The steps of the automatic population are:
+
+1. Generate a list of email address candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of phone number candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of username candidates from the identities, candidates from newer identities are ordered first.
+1. Generate a list of given name and family name candidates from the identities, candidates from newer identities are ordered first.
+1. If the `email` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `phone_number` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `preferred_username` standard attribute does not refer to a candidate in the list, clear it.
+1. If the `email` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If the `phone_number` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If the `preferred_username` standard attribute is absent and the list of candidate is non-empty, set it to the first candidate in the list.
+1. If both the `given_name` and the `family_name` standard attribute are absent and the list of candidate is non-empty, set it to the first candidate in the list.
+
+## Custom attributes
+
+In addition to standard attributes, the developer can define their own custom attributes.
+
+### Defining custom attributes
+
+The custom attributes are defined as a single JSON schema written against a subset of JSON schema 2019-09.
+
+Here is an example of the schema of the custom attributes.
+
+```JSON
+{
+  "properties": {
+    "app_user_role": {
+      "type": "string",
+      "enum": ["owner", "editor", "viewer"]
+    },
+    "stripe_customer_id": {
+      "type": "string"
+    }
+  }
+}
+```
+
+All changes made to custom attributes must be valid against the schema.
+
+#### Supported subset of JSON schema 2019-09
+
+- `type`: `boolean`, `string`, `number`, `integer`
+- `enum`
+- `multipleOf`
+- `maximum`
+- `exclusiveMaximum`
+- `minimum`
+- `exclusiveMinimum`
+- `maxLength`
+- `minLength`
+- `properties`
+
+### Custom attributes and the resolver
```

Comment:
**Discussion:** Should standard attributes also be included in the header of the resolver response?
Here we assume the users use custom attr for access control, returning them in the header will be helpful coz they dont need to call admin api in every request.